### PR TITLE
Add setup.py that builds Python extension

### LINKED ISSUE #630: Update `setup.py` to build c module (`pip install .`)
## What

`pip install .` should produce a fully working python package (including c components) without additional steps.

## Why

This is a pre-requisite for packaging for distribution, and is an expectation from a project with a `setup.py` file.

## Acceptance Criteria

- `pip install .` installs a fully working `openassetio` python package as per standard `pip` behaviour.
- python cmodule is static linked to dependencies
- No additional build steps are required aside from, in the first instance, assuming the starting condition that:
  - build dependencies are satisfied in the current environment (ie: cmake, conan, compiler)
  - cmake find package works to the "desired" dependencies for the user (eg: use toolchain ENV, or ...)
  - Instructions can include simple copy-paste one liners for happy path to achieve the above with conan central or similar.
- additional integration test that pip-installs the result and runs pytest test
  - shouldn't run with existing pytest/ctest tests
  - should be runnable by self-contained means without knowing "magic steps" (cross-platform) (eg. tox, py script).
   -  makes a new venv from build interpreter, installs using `pip install .`, runs pytest against that venv.
- BUILDING.md and other developer guidelines updated
- Follow the `pyproject.toml`/non legacy route described [here](https://pip.pypa.io/en/stable/reference/build-system/).

## Out of scope

- Significant build doc rewrites other than fixing up existing commands that are no longer relevant
- Significant customization of `setup.py` cmake invocation
- Adjusting existing test infrastructure invocation etc..
- Improvements to Window build process (eg: avoiding the need for `vcvars64.bat` or similar).

### PR DESCRIPTION
**Pull request by feltech**
Adds a `setup.py` that invokes cmake to install a complete/working `openassetio` python module from source via pip, for python-only integrations.

> **Warning**
We haven't updated the build instructions yet (see #624). At present, `setup.py` requires the C++ build toolchain to work in your environment (ie. cmake installed, and able to find dependencies + compiler).  See [BUILDING.md](BUILDING.md) and [CI bootstrap scripts](resources/build) for tips. 

Once these dependencies are satisfied, then build + installation should only require `python -m pip install .`.

Closes #630

### CODE REVIEW

**Comment by feltech** [comment]
> We haven't updated the build instructions yet

I did update the README a little to cover this change, but the updates got lost somewhere (I suspect an errant `git reset --hard`).  I've recovered those changes now (including RELEASE_NOTES) and squashed them in.

**Comment by feltech** [comment]
I removed `get_outputs` and `get_output_mapping` overrides from `setup.py` (squashed out of existence).  I put them in originally because during my investigations of OTIO's `setup.py` I found those overrides were necessary to support `--editable` installs.  However,  it turns out to be unnecessary for us!

File: pyproject.toml
Comment (JeanChristopheMorinPerso) [inline pyproject.toml:6]: I would personally just put a minimum version instead of pinning completely that dependency, `setuptools>=65.5.0`.

**Review by JeanChristopheMorinPerso — COMMENTED** [review]
This look really nice and clean for a first pass! I left some notes in the review.

Additional notes:

* I tested the PR and most things look good! I'm pretty surprised that the editable install works out of the box without any customization in the `setup.py`. Thee is some black magic going on that I'll need to investigate more to understand (for my own curiosity).
* I did not test on Windows and macOs yet.
* You will need a [MANIFEST.in](https://packaging.python.org/en/latest/guides/using-manifest-in/) file. The manifest file is used to
specify what to include in an sdist (source distribution). sdists
are distributed on PyPI alongside the wheels to let users
build a library from source when a matching wheel is not found or
if the user does something like `pip install openassetio --no-binary :all:`.
In this case, pip would download the sdist, extract it and compile+install
the project from source.

   By adding `conan` to your build requirements and running the `conan install`
command from the setup.py, you'll be able to have a fully self contained install.
Note that the comment about not install cpython still applies for this case.
Basically, there is no need to install cpython because when then `setup.py` file is
being executed, Python is already available (obviously).
* The `setup.py` is missing a lot of metadata (author, project url, license, description, etc). See https://packaging.python.org/en/latest/guides/distributing-packages-using-setuptools/#setup-args.

File: resources/build/conanfile.py
Comment (feltech) [reply resources/build/conanfile.py:25]: Thanks for the detailed explanation! 

We have a parallel dev stream going on looking into building wheels on CI, where we have indeed used `cibuildwheel` and disabled the conan `cpython` requirement in that case, so that the external Python environment is used instead. We'll hopefully have a PR up for that within the next week or so.

Really interesting note on into the glibc version - I hadn't considered that.

**Comment by feltech** [comment]
> * I did not test on Windows and macOs yet.

We're still skipping macOS for now, due to a Conan issue with the PyBind package on macOS: https://github.com/conan-io/conan-center-index/pull/11075 :disappointed: 

> * You will need a [MANIFEST.in](https://packaging.python.org/en/latest/guides/using-manifest-in/) file. The manifest file is used to
>   specify what to include in an sdist (source distribution).

This is something we hadn't considered at all! Thanks for pointing this out.
 
> * The `setup.py` is missing a lot of metadata (author, project url, license, description, etc). See https://packaging.python.org/en/latest/guides/distributing-packages-using-setuptools/#setup-args.

Good spot, thanks for the link.

I'm not sure how much will be added to this PR or if we'll defer to be part of the PyPI work we're doing.

File: setup.py
Comment (feltech) [reply setup.py]: Fixed in 0fb1f91a, thanks for the spot.

**Comment by feltech** [comment]
> * The `setup.py` is missing a lot of metadata

Added in 0fb1f91a

**Comment by elliotcmorris** [comment]
Do you think there's anything to be said for automatically trying to remove the build folder? I've come into the problem during dev of having to manually remove it a fair amount, but aware that may be a problem that doesn't become relevant to non contributors.

**Comment by feltech** [comment]
> Do you think there's anything to be said for automatically trying to remove the build folder? I've come into the problem during dev of having to manually remove it a fair amount, but aware that may be a problem that doesn't become relevant to non contributors.

Yeah it's annoying and non-obvious (I'm thinking of the "cannot find ninja" type errors).  But I'd hesitate to assume we have the right to remove it automatically, and I'm not sure where we would add the code to do it.  

Worth noting it's not a problem on CI, since it's a one-shot thing so never tries to re-use the `build` folder.

Maybe it's worth a separate issue to investigate why we need to remove the `build` directory in the first place?

**Comment by antirotor** [comment]
Just general notes I've hit trying to run `python -m build --wheel` - I always hit missing pybind11 error:

```
CMake Error at cmake/ThirdParty.cmake:55 (find_package):
  By not providing "Findpybind11.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "pybind11",
  but CMake did not find one.
```

**Comment by JeanChristopheMorinPerso** [comment]
@antirotor that's because right now the process of running `conan install` and actually building the wheel is a seperate process and require to set `CMAKE_TOOLCHAIN_FILE` to `~/conan/conan_paths.cmake` like it's done at https://github.com/OpenAssetIO/OpenAssetIO/pull/663/files#diff-c26c510aacd5c04c95ead8fd5f471a970232b297bc192cd445fef7e20365835cR62.

Moving the install step into the `setup.py` should fix this.

**Comment by JeanChristopheMorinPerso** [comment]
@feltech I see you didn't set the `author`/`maintainer` and `author_email`/`maintainer_email` (note that it's often redundant to use both author and maintainer at the same time). I would suggest to add them even if it seems redundant. It doesn't hurt and it doesn't force users to use that email. It's just another way of contacting the author and also clarify ownership.

**Comment by feltech** [comment]
> I see you didn't set the `author`/`maintainer` and `author_email`/`maintainer_email`

The `author` field is there in 8580d318 (apologies, linked the wrong commit under the review comment).

I didn't add the `author_email`, reasoning in the commit message

> The `author_email` field has been skipped, for fear of spam, and since
> its unclear who this should point to. It should also hopefully encourage
> people to engage out in the open (i.e. via Github).

Perhaps we should re-evaluate :thinking:

File: setup.py
Comment (JeanChristopheMorinPerso) [inline setup.py:105]: I'm no expert on this, but should this be the OpenAssetIO community or something like that since OpenAssetIO is an ASWF project?

**Comment by JeanChristopheMorinPerso** [comment]
> The author field is there in https://github.com/OpenAssetIO/OpenAssetIO/commit/8580d318519e0fe2475036dc16d8292de722c6d9 (apologies, linked the wrong commit under the review comment).
> 
> I didn't add the author_email, reasoning in the commit message

Ah, I see. I looked too quickly. Sorry about that. I think I would still add an email just as a last resort and for consistency.

BTW, the email won't show up in the PyPI UI. It's just metadata. The UI will show something like this:

Screen Shot 2022-10-20 at 9 29 43 AM

File: setup.py
Comment (feltech) [reply setup.py:105]: Good point, I suspect you're correct. I'll have to check on that.

File: setup.py
Comment (feltech) [inline setup.py:82]: [quesion] On [further reading](https://packaging.python.org/en/latest/tutorials/packaging-projects/#creating-pyproject-toml) I wonder if we should bundle as much metadata as possible in `pyproject.toml` to do things the "modern" way?

File: setup.py
Comment (feltech) [inline setup.py:81]: [question] This could be specified automatically if we use [setuptools_scm](https://github.com/pypa/setuptools_scm), which also handles bundling the repo for `sdist` packages. 

We may need to conform our git tag naming convention to [PEP 440](https://peps.python.org/pep-0440/#pre-releases) for that to work properly, though. And if we want to explicitly exclude things from the `sdist` we'd need a `MANIFEST.in` regardless, which would remove some of the appeal.

On balance I think the forcing Python package version to follow the git tag is likely to cause some pain for not much gain. But worth throwing the question out there.

File: setup.py
Comment (JeanChristopheMorinPerso) [reply setup.py:82]: It would indeed be a good idea.

File: setup.py
Comment (JeanChristopheMorinPerso) [reply setup.py:81]: I like my sdists to be leaner than the repository itself, meaning only the files required for building the project + tests. But that's going to be up to the OpenAssetIO TSC to decide if you want to use `setuptools_scm` or not.

File: setup.py
Comment (foundrytom) [inline setup.py:76]: [question] What is the reason for re-mapping `stderr` here?

File: README.md
Comment (foundrytom) [inline README.md:174]: [consider] Given we're going this far, until we sort #624, is it worth just putting the few lines of `conan install` here instead of more general toolchain details? This has been unfamiliar turf for many attempting to get going with python only builds.

File: src/openassetio-core/CMakeLists.txt
Comment (foundrytom) [inline src/openassetio-core/CMakeLists.txt:44]: [consider] Splitting these build changes into a separate PR (or the previous one?) as this one seems CI related otherwise.

File: resources/build/conanfile.py
Comment (foundrytom) [reply resources/build/conanfile.py:29]: Capturing off-line discussion I (think) we settled on using an env var to avoid changing the `package_id`

File: setup.py
Comment (foundrytom) [reply setup.py:81]: > We may need to conform our git tag naming convention to [PEP 440](https://peps.python.org/pep-0440/#pre-releases) for that to work properly, though. 

We'd originally followed the examples on [semver.org](https://semver.org/#spec-item-9) for the alpha pre-releases, but it looks like `a` is just as valid as `alpha`, so no harm switching moving forwards if it makes life more consistent.

File: setup.py
Comment (foundrytom) [inline setup.py:105]: [fix]
```suggestion
    author="Contributors to the OpenAssetIO project",
```

**Comment by foundrytom** [comment]
> @feltech I see you didn't set the `author`/`maintainer` and `author_email`/`maintainer_email` (note that it's often redundant to use both author and maintainer at the same time). I would suggest to add them even if it seems redundant. It doesn't hurt and it doesn't force users to use that email. It's just another way of contacting the author and also clarify ownership.

@feltech `openassetio-discussion@lists.aswf.io`

File: setup.py
Comment (foundrytom) [inline setup.py:68]: [fix] Can we stick with 3.7 compatible type hints to save having to revisit this?

**Review by foundrytom — CHANGES_REQUESTED** [review]
This is epic!

```
export CMAKE_TOOLCHAIN_FILE=~/.conan/conan_paths.cmake   
pip install .
```

Worked a charm. Just a few fairly minor tweaks.

**Comment by foundrytom** [comment]
@feltech The CI refactor has upset the required checks linking, we can fix that once this is merged.

File: resources/build/conanfile.py
Comment (feltech) [reply resources/build/conanfile.py:20]: Added passing `Python_EXECUTABLE` CMake variable to `setup.py` in 8aaeef81.  Not technically necessary for this PR, but good practice.

File: pyproject.toml
Comment (feltech) [reply pyproject.toml]: Good shout, updated to use a minimum version range in d1496af5

File: resources/build/conanfile.py
Comment (feltech) [reply resources/build/conanfile.py]: Simply removed the `options` entry and associated conditional in b260bbc2

File: setup.py
Comment (feltech) [reply setup.py]: Moved all the metadata to `pyproject.toml` in e9807bcd

File: setup.py
Comment (feltech) [reply setup.py]: We decided to add a follow-up ticket to deal with `sdist`s, one way or another: https://github.com/OpenAssetIO/OpenAssetIO/issues/666

File: setup.py
Comment (feltech) [reply setup.py:80]: When debugging build output I noticed some info was still being swallowed, I forget what now, but this fixed it.  I can try to reproduce, if you like, so I can add a comment?

Comment (foundrytom) [reply setup.py:80]: A comment could be good, so we don't forget why once we've slept.

**Comment by feltech** [comment]
> I see you didn't set the [...] `author_email`/`maintainer_email`

Added as part of the migration to `pyproject.toml` in e9807bcd (Tom can deal with any spam :smiling_imp: )

File: setup.py
Comment (feltech) [reply setup.py]: Ah I didn't get there quite fast enough before you started reviewing, already updated in 4161630a

File: setup.py
Comment (feltech) [reply setup.py]: I updated as part of the move to `pyproject.toml` in e9807bcd but I used "The OpenAssetIO Community", should I change that?

Comment (foundrytom) [reply setup.py]: This was just to be consistent with OTIO - whichever your prefer.

Comment (feltech) [reply setup.py]: Updated author name and email in af4ae75e

**Comment by feltech** [comment]
> > @feltech I see you didn't set the `author`/`maintainer` and `author_email`/`maintainer_email` (note that it's often redundant to use both author and maintainer at the same time). I would suggest to add them even if it seems redundant. It doesn't hurt and it doesn't force users to use that email. It's just another way of contacting the author and also clarify ownership.
> 
> @feltech `openassetio-discussion@lists.aswf.io`

Updated in af4ae75e (see also https://github.com/OpenAssetIO/OpenAssetIO/pull/663#discussion_r1001636498)

File: resources/build/conanfile.py
Comment (feltech) [reply resources/build/conanfile.py]: > we settled on using an env var

Apologies, that wasn't my take-away, should I update this?

Comment (foundrytom) [reply resources/build/conanfile.py]: We can sort later if its not needed - sorry thought you made the good point that it would speed up the wheel builds.

File: README.md
Comment (feltech) [reply README.md:184]: Added some additional instructions in a2cf809f, how does that look?

Comment (JeanChristopheMorinPerso) [reply README.md:184]: I'd argue that the Conan install should be executed from the setup.py if you want to make a good user experience and will anyway be required for sdists.

Comment (foundrytom) [reply README.md:184]: We're going to look into this, just in a separate task #666

Comment (foundrytom) [reply README.md:184]: @JeanChristopheMorinPerso Do you have any experiencing in making this work on Windows too - is it fair to expect that people are already in a developer command prompt?

File: src/openassetio-core/CMakeLists.txt
Comment (foundrytom) [reply src/openassetio-core/CMakeLists.txt:44]: Haha sorry yes.

File: README.md
Comment (JeanChristopheMorinPerso) [reply README.md:184]: Yep, it is entirely fair to expect that the user has a compiler available and that he is in a developer command prompt properly setup to compile C++ code. You do that same assumption on Linux and macOs, so it would be weird to make a different assumption on Windows.

File: .github/workflows/build-instructions.yml
Comment (JeanChristopheMorinPerso) [inline .github/workflows/build-instructions.yml:38]: Cannot CMake usually take care of sourcing the right compiler and environment?

File: setup.py
Comment (feltech) [reply setup.py]: Well, try as I might, analysing my bash history and file history and trying to reproduce the issue, I can't.  So just removed in b74038ca

File: .github/workflows/build-instructions.yml
Comment (feltech) [reply .github/workflows/build-instructions.yml:38]: Ah, it might be compounded by us using the Ninja generator - not so much CMake assuming MinGW, but Ninja.

Comment (JeanChristopheMorinPerso) [reply .github/workflows/build-instructions.yml:38]: Ah, that would make sense.

Comment (foundrytom) [reply .github/workflows/build-instructions.yml:38]: If there is any way we can remove this malarky that would be _amazing_ (maybe wrap this into #538?)

File: README.md
Comment (foundry-markf) [inline README.md:165]: Nit: As a developer, I almost always use `pip install -e .` so I can noodle on Python code. Granted, I do that in a pure Python environment.

Perhaps this needs a little clarification for intended usage?

File: .github/workflows/build-instructions.yml
Comment (feltech) [reply .github/workflows/build-instructions.yml:38]: I kinda like having Ninja as a cross-platform consistent generator, which as a bonus trivially parallelizes builds. But if it can completely remove the need for vcvarsall in CI, then removing it (or making it conditional on OS) is probably worthwhile.

Comment (JeanChristopheMorinPerso) [reply .github/workflows/build-instructions.yml:38]: Ninja is nice indeed as it speeds up builds on Windows. Maybe take a look at OCIO (which I think use Ninja too) and see how they do it?

Comment (feltech) [reply .github/workflows/build-instructions.yml:38]: > maybe wrap this into #538

That would get my vote, but happy to try it out in this PR if you like.

File: README.md
Comment (feltech) [reply README.md:165]: :thinking:  I think people who don't know about develop installs will be wanting to install into a wider system, so the docs are appropriate, whereas people who do know about develop installs will be able to understand that they can add `-e` if they want?

File: .github/workflows/build-instructions.yml
Comment (feltech) [reply .github/workflows/build-instructions.yml:38]: > Maybe take a look at OCIO

They skip Ninja on Windows (https://github.com/AcademySoftwareFoundation/OpenColorIO/blob/660cf515faa491fadab188eb0f2ed62b30b4b4f6/setup.py#L93) because

> MSVC would require all variables be exported for Ninja to pick it up, which is a little tricky to do

:laughing:

File: README.md
Comment (foundry-markf) [reply README.md:165]: True, although when I first read it (as a developer) I wondered why the develop mode wasn't being used, almost as if it wasn't supported. Perhaps a note on that?

File: .github/workflows/build-instructions.yml
Comment (foundrytom) [reply .github/workflows/build-instructions.yml:38]: Lets save this for another day

Comment (feltech) [reply .github/workflows/build-instructions.yml:38]: Added an investigation ticket to check feasibility of using the MSVC generator #669

File: README.md
Comment (feltech) [reply README.md:165]: :+1: added a note in 03d1a46b, is that OK?

Comment (foundry-markf) [reply README.md:165]: Thanks!

File: resources/build/conanfile.py
Comment (feltech) [reply resources/build/conanfile.py]: I've added a note on #654 to do this when we get around to creating a Conan package. Does that sound reasonable?

Comment (foundrytom) [reply resources/build/conanfile.py]: LGTM!

File: src/openassetio-core/CMakeLists.txt
Comment (feltech) [reply src/openassetio-core/CMakeLists.txt:44]: Moved the CMakeLists changes into the previous (i.e. first) commit e4fdb46e

**Pull request merged**
